lib/spack/spack/parser.py

```diff
@@ -467,6 +468,13 @@ def parse(
                 assert match, "SPLIT_KVP and PROPAGATED_KEY_VALUE_PAIR do not agree."
 
                 name, delim, value = match.groups()
+                if name in spack.directives.reserved_names:
+                    raise SpecParsingError(
+                        f"propagation is not supported for {name}",
+                        self.ctx.current_token,
+                        self.literal_str,
+                    )
```

Comment:
If you move this into `Spec._add_flag` then the changes to this module are probably all unnecessary.

Reply:
If we move it to `_add_flag`, it's much harder to produce a useful error message showing the spec, b/c we don't have the `literal_str` being parsed, or the token.

I guess I could have `_add_flag` raise something, and wrap `_add_flag` inside the spec parser to catch errors and raise a parsing error with the token, etc. Would that work?  That would at least put the checking logic in one place, and would potentially allow other code to raise errors and have them produce useful errors with nice underlining. Thoughts?

Reply:
Yeah I think that would work. We could also just add `asserts` for the other ones in `_add_flag` if that makes more sense to you, it just felt weird to have the rest accept it and this one doesn't.

Reply:
Ok this is done -- see what you think.